### PR TITLE
[SPARK-44736][CONNECT] Add Dataset.explode to Spark Connect Scala Client

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds Dataset.explode to the Spark Connect Scala Client.

### Why are the changes needed?
To increase compatibility with the existing Dataset API in sql/core.

### Does this PR introduce _any_ user-facing change?
Yes, it adds a new method to the scala client.

### How was this patch tested?
I added a test to `UserDefinedFunctionE2ETestSuite`.

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -508,7 +508,8 @@ class SparkConnectPlanner(val sessionHolder: SessionHolder) extends Logging {
     val commonUdf = rel.getFunc
     commonUdf.getFunctionCase match {
       case proto.CommonInlineUserDefinedFunction.FunctionCase.SCALAR_SCALA_UDF =>
-        transformTypedMapPartitions(commonUdf, baseRel)
+        val analyzed = session.sessionState.executePlan(baseRel).analyzed
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:511]: This was a bug. We only tested mapPartitions with resolved logical plans...

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2724,6 +2725,38 @@ class Dataset[T] private[sql] (
...
+   * Using `flatMap()` this can similarly be exploded as:
+   *
+   * {{{
+   *   ds.flatMap(_.words.split(" "))
+   * }}}
+   *
+   * @group untypedrel
+   * @since 3.5.0
+   */
+  @deprecated("use flatMap() or select() with functions.explode() instead", "2.0.0")
```
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2755]:
```suggestion
  @deprecated("use flatMap() or select() with functions.explode() instead", "3.5.0")
```

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -2724,6 +2725,38 @@ class Dataset[T] private[sql] (
...
+  def explode[A <: Product: TypeTag](input: Column*)(f: Row => TraversableOnce[A]): DataFrame = {
```
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2756]: Do we need `String` signature too?

**Review by zhenlineo — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

**Review by amaliujia — APPROVED** [review]

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 3 above)
Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2756]: I meant if we need to add the string signature version of this: `def explode[A, B : TypeTag](inputColumn: String, outputColumn: String)(f: A => TraversableOnce[B])` signature that exists in non Connect Dataset.

**Comment by hvanhovell** [comment]
Merging.

**Pull request closed without merging**